src/jdk.incubator.vector/share/classes/jdk/incubator/vector/DoubleVector.java

```diff
@@ -1725,11 +1725,48 @@ else if (op == IS_FINITE ||
      * {@inheritDoc} <!--workaround-->
      */
     @Override
-    @ForceInline
-    public final
+    public abstract
     VectorMask<Double> test(VectorOperators.Test op,
-                                  VectorMask<Double> m) {
-        return test(op).and(m);
+                                  VectorMask<Double> m);
+
+    /*package-private*/
+    @ForceInline
+    final
+    <M1 extends VectorMask<Double>,
+     M2 extends VectorMask<Long>>
+    M1 testTemplate(Class<M1> maskType, Test op, M2 mask) {
```

Comment:
Can we simplify by making some code FP specific?
- the mask `cast` can be applied in this method rather than in the caller, simplifying the signature
- for clarify, vector `viewAsIntegralLanes` is only needed for FP (update would be required for the non-mask template).